jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -376,65 +392,44 @@ private void clearArray(RetainedBucket[] poolArray, AtomicLong memoryCounter)
     private void releaseExcessMemory(boolean direct)
     {
         long maxMemory = direct ? _maxDirectMemory : _maxHeapMemory;
-        if (maxMemory > 0)
-        {
-            long excess = getMemory(direct) - maxMemory;
-            if (excess > 0)
-                evict(direct, excess);
-        }
-    }
+        if (maxMemory <= 0)
+            return;
 
-    /**
-     * This eviction mechanism searches for the RetainableByteBuffers that were released the longest time ago.
-     *
-     * @param direct true to search in the direct buffers buckets, false to search in the heap buffers buckets.
-     * @param excess the amount of bytes to evict. At least this much will be removed from the buckets.
-     */
-    private void evict(boolean direct, long excess)
-    {
-        if (LOG.isDebugEnabled())
-            LOG.debug("evicting {} bytes from {} pools", excess, (direct ? "direct" : "heap"));
-        long now = NanoTime.now();
-        long totalClearedCapacity = 0L;
+        if (_evictor.compareAndSet(false, true))
+        {
+            // we are the one and only thread evicting
```

Comment:
I'm not fond of this idea of having a single thread doing all the eviction work, as that could create some pretty bad latency outliers. IMHO every thread that figures we have excess memory should participate in the eviction process.

Reply:
Perhaps we limit the thread to just a single iteration to reduce. It then will never have a bad latency, but if it not enough, then the next release can free more.

Reply:
That would work, assuming we accept overshooting for some duration.